<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />
<meta name="viewport" content="width=device-width, initial-scale=1">
<meta property="og:title" content="Chapter 5 Data visualization | Introduction to bioinformatics" />
<meta property="og:type" content="book" />


<meta property="og:description" content="Course material for the Introduction to bioinformatics (WSBIM1207) course at UCLouvain." />


<meta name="author" content="Laurent Gatto" />

<meta name="date" content="2019-03-12" />


<meta name="description" content="Course material for the Introduction to bioinformatics (WSBIM1207) course at UCLouvain.">

<title>Chapter 5 Data visualization | Introduction to bioinformatics</title>

<link href="libs/tufte-css-2015.12.29/tufte.css" rel="stylesheet" />
<link href="libs/tufte-css-2015.12.29/envisioned.css" rel="stylesheet" />
<link href="libs/msmb-css-0/msmb.css" rel="stylesheet" />
<script>
function toggle_visibility(id1, id2) {
var e = document.getElementById(id1);
var f = document.getElementById(id2);

e.style.display = ((e.style.display!='none') ? 'none' : 'block');

if(f.classList.contains('fa-plus-square')) {
    f.classList.add('fa-minus-square')
    f.classList.remove('fa-plus-square')
} else {
    f.classList.add('fa-plus-square')
    f.classList.remove('fa-minus-square')
}

}
</script>
<script src="libs/htmlwidgets-1.3/htmlwidgets.js"></script>
<script src="libs/jquery-1.12.4/jquery.min.js"></script>
<link href="libs/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet" />
<script src="libs/datatables-binding-0.5/datatables.js"></script>
<link href="libs/dt-core-1.10.16/css/jquery.dataTables.min.css" rel="stylesheet" />
<link href="libs/dt-core-1.10.16/css/jquery.dataTables.extra.css" rel="stylesheet" />
<script src="libs/dt-core-1.10.16/js/jquery.dataTables.min.js"></script>
<link href="libs/crosstalk-1.0.0/css/crosstalk.css" rel="stylesheet" />
<script src="libs/crosstalk-1.0.0/js/crosstalk.min.js"></script>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>



</head>

<body>



<div class="row">
<div class="col-sm-12">
<div id="TOC">
<ul class="navbar">
<li class="msmb"><p class="title">Introduction to bioinformatics<p><p class="author">Laurent Gatto</p>
<li class="dropdown" style="float:right">
<a href="javascript:void(0)" class="dropbtn">&#x25BE; Chapters</a>
<div class="dropdown-content">
<a href="index.html">Preamble</a>
<a href="sec-dataorg.html"><span class="toc-section-number">1</span> Data organisation with Spreadsheets</a>
<a href="sec-startr.html"><span class="toc-section-number">2</span> Starting with R</a>
<a href="sec-startdata.html"><span class="toc-section-number">3</span> Starting with data</a>
<a href="sec-dplyr.html"><span class="toc-section-number">4</span> Manipulating and analyzing data with dplyr</a>
<a id="active-page" href="sec-vis.html"><span class="toc-section-number">5</span> Data visualization</a><ul class="toc-sections">
<li class="toc"><a href="#plotting-with-ggplot2"> Plotting with <strong><code>ggplot2</code></strong></a></li>
<li class="toc"><a href="#building-your-plots-iteratively"> Building your plots iteratively</a></li>
<li class="toc"><a href="#boxplot"> Boxplot</a></li>
<li class="toc"><a href="#plotting-time-series-data"> Plotting time series data</a></li>
<li class="toc"><a href="#faceting"> Faceting</a></li>
<li class="toc"><a href="#ggplot2-themes"> <strong><code>ggplot2</code></strong> themes</a></li>
<li class="toc"><a href="#customization"> Customization</a></li>
<li class="toc"><a href="#arranging-and-exporting-plots"> Arranging and exporting plots</a></li>
<li class="toc"><a href="#other-packages-for-visualisation"> Other packages for visualisation</a></li>
<li class="toc"><a href="#additional-exercises-4"> Additional exercises</a></li>
</ul>
<a href="sec-join.html"><span class="toc-section-number">6</span> Joining tables</a>
<a href="sec-rr.html"><span class="toc-section-number">7</span> Reproducible research</a>
<a href="sec-bioinfo.html"><span class="toc-section-number">8</span> Bioinformatics</a>
<a href="sec-prog.html"><span class="toc-section-number">9</span> Additional programming concepts</a>
<a href="sec-ccl.html"><span class="toc-section-number">10</span> Conclusions</a>
<a href="session-information.html">Session information</a>
</div>
</li>
</ul>
</div>
</div>
</div>
<div class="row">
<div class="col-sm-12">
<div id="sec:vis" class="section level1">
<h1><span class="header-section-number">Chapter 5</span> Data visualization</h1>
<p><strong>Learning Objectives</strong></p>
<ul>
<li>Produce scatter plots, boxplots, and time series plots using ggplot.</li>
<li>Set universal plot settings.</li>
<li>Describe what faceting is and apply faceting in ggplot.</li>
<li>Modify the aesthetics of an existing ggplot plot (including axis labels and color).</li>
<li>Build complex and customized plots from data in a data frame.</li>
</ul>
<p>We start by loading the required packages. <strong><code>ggplot2</code></strong> is included in the <strong><code>tidyverse</code></strong> package.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(<span class="st">&quot;tidyverse&quot;</span>)</code></pre></div>
<p>If not still in the workspace, load the data we saved in the previous lesson.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">surveys_complete &lt;-<span class="st"> </span><span class="kw">read_csv</span>(<span class="st">&quot;data_output/surveys_complete.csv&quot;</span>)</code></pre></div>
<p>The <a href="https://github.com/rstudio/cheatsheets/raw/master/data-visualization-2.1.pdf">Data Visualization Cheat Sheet</a> will cover the basics and more advanced features of <code>ggplot2</code> and will help, in addition to serve as a reminder, getting an overview of the many data representations available in the package.</p>
<div id="plotting-with-ggplot2" class="section level2">
<h2><span class="header-section-number">5.1</span> Plotting with <strong><code>ggplot2</code></strong></h2>
<p><strong><code>ggplot2</code></strong> is a plotting package that makes it simple to create complex plots from data in a data frame. It provides a more programmatic interface for specifying what variables to plot, how they are displayed, and general visual properties. The theoretical foundation that supports the <code>ggplot2</code> is the <em>Grammar of Graphics</em> (<span class="citation">Wilkinson (<label for="tufte-mn-19" class="margin-toggle">&#8853;</label><input type="checkbox" id="tufte-mn-19" class="margin-toggle">2005<span class="marginnote">Wilkinson, Leland. 2005. <em>The Grammar of Graphics (Statistics and Computing)</em>. Berlin, Heidelberg: Springer-Verlag.</span>)</span>). Using this approach, we only need minimal changes if the underlying data change or if we decide to change from a bar plot to a scatterplot. This helps in creating publication quality plots with minimal amounts of adjustments and tweaking.</p>
<p>There is a book about <code>ggplot2</code> (<span class="citation">Wickham (<label for="tufte-mn-20" class="margin-toggle">&#8853;</label><input type="checkbox" id="tufte-mn-20" class="margin-toggle">2016<span class="marginnote">Wickham, Hadley. 2016. <em>ggplot2: Elegant Graphics for Data Analysis</em>. Springer-Verlag New York. <a href="http://ggplot2.org" class="uri">http://ggplot2.org</a>.</span>)</span>) that provides a good overview, but it is outdated. The <code>ggplot2</code> webpage (<a href="https://ggplot2.tidyverse.org" class="uri">https://ggplot2.tidyverse.org</a>) provides ample documentation.</p>
<p><code>ggplot2</code> functions like data in the ‘long’ format, i.e., a column for every dimension, and a row for every observation. Well-structured data will save you lots of time when making figures with <code>ggplot2</code>.</p>
<p>ggplot graphics are built step by step by adding new elements. Adding layers in this fashion allows for extensive flexibility and customization of plots.</p>
<p>To build a ggplot, we will use the following basic template that can be used for different types of plots:</p>
<pre><code>ggplot(data = &lt;DATA&gt;, mapping = aes(&lt;MAPPINGS&gt;)) +  &lt;GEOM_FUNCTION&gt;()</code></pre>
<ul>
<li>use the <code>ggplot()</code> function and bind the plot to a specific data frame using the <code>data</code> argument</li>
</ul>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">ggplot</span>(<span class="dt">data =</span> surveys_complete)</code></pre></div>
<ul>
<li>define a mapping (using the aesthetic (<code>aes</code>) function), by selecting the variables to be plotted and specifying how to present them in the graph, e.g. as x/y positions or characteristics such as size, shape, color, etc.</li>
</ul>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">ggplot</span>(<span class="dt">data =</span> surveys_complete, <span class="dt">mapping =</span> <span class="kw">aes</span>(<span class="dt">x =</span> weight, <span class="dt">y =</span> hindfoot_length))</code></pre></div>
<ul>
<li><p>add ‘geoms’ – graphical representations of the data in the plot (points, lines, bars). <strong><code>ggplot2</code></strong> offers many different geoms; we will use some common ones today, including:</p>
<ul>
<li><code>geom_point()</code> for scatter plots, dot plots, etc.</li>
<li><code>geom_boxplot()</code> for, well, boxplots!</li>
<li><code>geom_line()</code> for trend lines, time series, etc.</li>
</ul></li>
</ul>
<p>To add a geom to the plot use the <code>+</code> operator. Because we have two continuous variables, let’s use <code>geom_point()</code> first:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">ggplot</span>(<span class="dt">data =</span> surveys_complete, <span class="dt">mapping =</span> <span class="kw">aes</span>(<span class="dt">x =</span> weight, <span class="dt">y =</span> hindfoot_length)) <span class="op">+</span>
<span class="st">  </span><span class="kw">geom_point</span>()</code></pre></div>
<p><img src="WSBIM1207_files/figure-html/first-ggplot-1.png" width="672"  /></p>
<p>The <code>+</code> in the <strong><code>ggplot2</code></strong> package is particularly useful because it allows you to modify existing <code>ggplot</code> objects. This means you can easily set up plot templates and conveniently explore different types of plots, so the above plot can also be generated with code like this:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Assign plot to a variable</span>
surveys_plot &lt;-<span class="st"> </span><span class="kw">ggplot</span>(<span class="dt">data =</span> surveys_complete,
                       <span class="dt">mapping =</span> <span class="kw">aes</span>(<span class="dt">x =</span> weight, <span class="dt">y =</span> hindfoot_length))

<span class="co"># Draw the plot</span>
surveys_plot <span class="op">+</span>
<span class="st">    </span><span class="kw">geom_point</span>()</code></pre></div>
<p><strong>Notes</strong></p>
<ul>
<li>Anything you put in the <code>ggplot()</code> function can be seen by any geom layers that you add (i.e., these are universal plot settings). This includes the x- and y-axis mapping you set up in <code>aes()</code>.</li>
<li>You can also specify mappings for a given geom independently of the mappings defined globally in the <code>ggplot()</code> function.</li>
<li>The <code>+</code> sign used to add new layers must be placed at the end of the line containing the <em>previous</em> layer. If, instead, the <code>+</code> sign is added at the beginning of the line containing the new layer, <strong><code>ggplot2</code></strong> will not add the new layer and will return an error message.</li>
</ul>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># This is the correct syntax for adding layers</span>
surveys_plot <span class="op">+</span>
<span class="st">  </span><span class="kw">geom_point</span>()

<span class="co"># This will not add the new layer and will return an error message</span>
surveys_plot
  <span class="op">+</span><span class="st"> </span><span class="kw">geom_point</span>()</code></pre></div>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Scatter plots can be useful exploratory tools for small datasets. For data sets with large numbers of observations, such as the <code>surveys_complete</code> data set, overplotting of points can be a limitation of scatter plots. One strategy for handling such settings is to use hexagonal binning of observations. The plot space is tessellated into hexagons. Each hexagon is assigned a color based on the number of observations that fall within its boundaries. To use hexagonal binning <code>ggplot2</code>, first install the R package <code>hexbin</code> from CRAN and load it.</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="solution">
<p class="solution-begin">
► Solution<span id="sol-start-37" class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility(&#39;sol-body-37&#39;, &#39;sol-start-37&#39;)"></span>
</p>
<div id="sol-body-37" class="solution-body" style="display: none;">
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">install.packages</span>(<span class="st">&quot;hexbin&quot;</span>)
<span class="kw">library</span>(<span class="st">&quot;hexbin&quot;</span>)</code></pre></div>
<p class="solution-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Then use the <code>geom_hex()</code> function</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">surveys_plot <span class="op">+</span>
<span class="st"> </span><span class="kw">geom_hex</span>()</code></pre></div>
<p>What are the relative strengths and weaknesses of a hexagonal bin plot compared to a scatter plot? Examine the above scatter plot and compare it with the hexagonal bin plot that you created.</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
</div>
<div id="building-your-plots-iteratively" class="section level2">
<h2><span class="header-section-number">5.2</span> Building your plots iteratively</h2>
<p>Building plots with <strong><code>ggplot2</code></strong> is typically an iterative process. We start by defining the dataset we’ll use, lay out the axes, and choose a geom:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">ggplot</span>(<span class="dt">data =</span> surveys_complete, <span class="dt">mapping =</span> <span class="kw">aes</span>(<span class="dt">x =</span> weight, <span class="dt">y =</span> hindfoot_length)) <span class="op">+</span>
<span class="st">    </span><span class="kw">geom_point</span>()</code></pre></div>
<p><img src="WSBIM1207_files/figure-html/create-ggplot-object-1.png" width="672"  /></p>
<p>Then, we start modifying this plot to extract more information from it. For instance, we can add transparency (<code>alpha</code>) to avoid overplotting:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">ggplot</span>(<span class="dt">data =</span> surveys_complete, <span class="dt">mapping =</span> <span class="kw">aes</span>(<span class="dt">x =</span> weight, <span class="dt">y =</span> hindfoot_length)) <span class="op">+</span>
<span class="st">    </span><span class="kw">geom_point</span>(<span class="dt">alpha =</span> <span class="fl">0.1</span>)</code></pre></div>
<p><img src="WSBIM1207_files/figure-html/adding-transparency-1.png" width="672"  /></p>
<p>We can also add colors for all the points:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">ggplot</span>(<span class="dt">data =</span> surveys_complete, <span class="dt">mapping =</span> <span class="kw">aes</span>(<span class="dt">x =</span> weight, <span class="dt">y =</span> hindfoot_length)) <span class="op">+</span>
<span class="st">    </span><span class="kw">geom_point</span>(<span class="dt">alpha =</span> <span class="fl">0.1</span>, <span class="dt">color =</span> <span class="st">&quot;blue&quot;</span>)</code></pre></div>
<p><img src="WSBIM1207_files/figure-html/adding-colors-1.png" width="672"  /></p>
<p>Or to color each species in the plot differently, you could use a vector as an input to the argument <strong>color</strong>. <strong><code>ggplot2</code></strong> will provide a different color corresponding to different values in the vector. Here is an example where we color with <strong><code>species_id</code></strong>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">ggplot</span>(<span class="dt">data =</span> surveys_complete, <span class="dt">mapping =</span> <span class="kw">aes</span>(<span class="dt">x =</span> weight, <span class="dt">y =</span> hindfoot_length)) <span class="op">+</span>
<span class="st">    </span><span class="kw">geom_point</span>(<span class="dt">alpha =</span> <span class="fl">0.1</span>, <span class="kw">aes</span>(<span class="dt">color =</span> species_id))</code></pre></div>
<p><img src="WSBIM1207_files/figure-html/color-by-species1-1.png" width="672"  /></p>
<p>We can also specify the colors directly inside the mapping provided in the <code>ggplot()</code> function. This will be seen by any geom layers and the mapping will be determined by the x- and y-axis set up in <code>aes()</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">ggplot</span>(<span class="dt">data =</span> surveys_complete, <span class="dt">mapping =</span> <span class="kw">aes</span>(<span class="dt">x =</span> weight, <span class="dt">y =</span> hindfoot_length, <span class="dt">color =</span> species_id)) <span class="op">+</span>
<span class="st">    </span><span class="kw">geom_point</span>(<span class="dt">alpha =</span> <span class="fl">0.1</span>)</code></pre></div>
<p><img src="WSBIM1207_files/figure-html/color-by-species2-1.png" width="672"  /></p>
<p>Notice that we can change the geom layer and colors will be still determined by <strong><code>species_id</code></strong></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">ggplot</span>(<span class="dt">data =</span> surveys_complete, <span class="dt">mapping =</span> <span class="kw">aes</span>(<span class="dt">x =</span> weight, <span class="dt">y =</span> hindfoot_length, <span class="dt">color =</span> species_id)) <span class="op">+</span>
<span class="st">    </span><span class="kw">geom_jitter</span>(<span class="dt">alpha =</span> <span class="fl">0.1</span>)</code></pre></div>
<p><img src="WSBIM1207_files/figure-html/color-by-species3-1.png" width="672"  /></p>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Use what you just learned to create a scatter plot of <code>weight</code> over <code>species_id</code> with the plot types showing in different colors. Is this a good way to show this type of data?</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="solution">
<p class="solution-begin">
► Solution<span id="sol-start-38" class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility(&#39;sol-body-38&#39;, &#39;sol-start-38&#39;)"></span>
</p>
<div id="sol-body-38" class="solution-body" style="display: none;">
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">ggplot</span>(<span class="dt">data =</span> surveys_complete, <span class="dt">mapping =</span> <span class="kw">aes</span>(<span class="dt">x =</span> species_id, <span class="dt">y =</span> weight)) <span class="op">+</span>
<span class="st">    </span><span class="kw">geom_point</span>(<span class="kw">aes</span>(<span class="dt">color =</span> plot_type))</code></pre></div>
<p><img src="WSBIM1207_files/figure-html/unnamed-chunk-177-1.png" width="672"  /></p>
<p class="solution-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
</div>
<div id="boxplot" class="section level2">
<h2><span class="header-section-number">5.3</span> Boxplot</h2>
<p>We can use boxplots to visualize the distribution of weight within each species:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">ggplot</span>(<span class="dt">data =</span> surveys_complete, <span class="dt">mapping =</span> <span class="kw">aes</span>(<span class="dt">x =</span> species_id, <span class="dt">y =</span> weight)) <span class="op">+</span>
<span class="st">    </span><span class="kw">geom_boxplot</span>()</code></pre></div>
<p><img src="WSBIM1207_files/figure-html/boxplot-1.png" width="672"  /></p>
<p>By adding points to boxplot, we can have a better idea of the number of measurements and of their distribution:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">ggplot</span>(<span class="dt">data =</span> surveys_complete, <span class="dt">mapping =</span> <span class="kw">aes</span>(<span class="dt">x =</span> species_id, <span class="dt">y =</span> weight)) <span class="op">+</span>
<span class="st">    </span><span class="kw">geom_boxplot</span>(<span class="dt">alpha =</span> <span class="dv">0</span>) <span class="op">+</span>
<span class="st">    </span><span class="kw">geom_jitter</span>(<span class="dt">alpha =</span> <span class="fl">0.3</span>, <span class="dt">color =</span> <span class="st">&quot;tomato&quot;</span>)</code></pre></div>
<p><img src="WSBIM1207_files/figure-html/boxplot-with-points-1.png" width="672"  /></p>
<p>Notice how the boxplot layer is behind the jitter layer? What do you need to change in the code to put the boxplot in front of the points such that it’s not hidden?</p>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Boxplots are useful summaries, but hide the <em>shape</em> of the distribution. For example, if the distribution is bimodal, we would not see it in a boxplot. An alternative to the boxplot is the violin plot, where the shape (of the density of points) is drawn.</p>
<ul>
<li>Replace the box plot with a violin plot; see <code>geom_violin()</code>.</li>
</ul>
<p>In many types of data, it is important to consider the <em>scale</em> of the observations. For example, it may be worth changing the scale of the axis to better distribute the observations in the space of the plot. Changing the scale of the axes is done similarly to adding/modifying other components (i.e., by incrementally adding commands). Try making these modifications:</p>
<ul>
<li>Represent weight on the log<sub>10</sub> scale; see <code>scale_y_log10()</code>.</li>
</ul>
<p>So far, we’ve looked at the distribution of weight within species. Try making a new plot to explore the distribution of another variable within each species.</p>
<ul>
<li><p>Create a boxplot for <code>hindfoot_length</code>. Overlay the boxplot layer on a jitter layer to show actual measurements.</p></li>
<li><p>Add color to the data points on your boxplot according to the plot from which the sample was taken (<code>plot_id</code>).</p></li>
</ul>
<p><em>Hint:</em> Check the class for <code>plot_id</code>. Consider changing the class of <code>plot_id</code> from integer to factor. Why does this change how R makes the graph?</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
</div>
<div id="plotting-time-series-data" class="section level2">
<h2><span class="header-section-number">5.4</span> Plotting time series data</h2>
<p>Let’s calculate number of counts per year for each species. First we need to group the data and count records within each group:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">yearly_counts &lt;-<span class="st"> </span>surveys_complete <span class="op">%&gt;%</span>
<span class="st">                 </span><span class="kw">count</span>(year, species_id)</code></pre></div>
<p>Time series data can be visualized as a line plot with years on the x axis and counts on the y axis:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">ggplot</span>(<span class="dt">data =</span> yearly_counts, <span class="dt">mapping =</span> <span class="kw">aes</span>(<span class="dt">x =</span> year, <span class="dt">y =</span> n)) <span class="op">+</span>
<span class="st">     </span><span class="kw">geom_line</span>()</code></pre></div>
<p><img src="WSBIM1207_files/figure-html/first-time-series-1.png" width="672"  /></p>
<p>Unfortunately, this does not work because we plotted data for all the species together. We need to tell ggplot to draw a line for each species by modifying the aesthetic function to include <code>group = species_id</code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">ggplot</span>(<span class="dt">data =</span> yearly_counts, <span class="dt">mapping =</span> <span class="kw">aes</span>(<span class="dt">x =</span> year, <span class="dt">y =</span> n, <span class="dt">group =</span> species_id)) <span class="op">+</span>
<span class="st">    </span><span class="kw">geom_line</span>()</code></pre></div>
<p><img src="WSBIM1207_files/figure-html/time-series-by-species-1.png" width="672"  /></p>
<p>We will be able to distinguish species in the plot if we add colors (using <code>color</code> also automatically groups the data):</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">ggplot</span>(<span class="dt">data =</span> yearly_counts, <span class="dt">mapping =</span> <span class="kw">aes</span>(<span class="dt">x =</span> year, <span class="dt">y =</span> n, <span class="dt">color =</span> species_id)) <span class="op">+</span>
<span class="st">    </span><span class="kw">geom_line</span>()</code></pre></div>
<p><img src="WSBIM1207_files/figure-html/time-series-with-colors-1.png" width="672"  /></p>
</div>
<div id="faceting" class="section level2">
<h2><span class="header-section-number">5.5</span> Faceting</h2>
<p><strong><code>ggplot2</code></strong> has a special technique called <em>faceting</em> that allows the user to split one plot into multiple plots based on a factor included in the dataset. We will use it to make a time series plot for each species:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">ggplot</span>(<span class="dt">data =</span> yearly_counts, <span class="dt">mapping =</span> <span class="kw">aes</span>(<span class="dt">x =</span> year, <span class="dt">y =</span> n)) <span class="op">+</span>
<span class="st">    </span><span class="kw">geom_line</span>() <span class="op">+</span>
<span class="st">    </span><span class="kw">facet_wrap</span>(<span class="op">~</span><span class="st"> </span>species_id)</code></pre></div>
<p><img src="WSBIM1207_files/figure-html/first-facet-1.png" width="672"  /></p>
<p>Now we would like to split the line in each plot by the sex of each individual measured. To do that we need to make counts in the data frame grouped by <code>year</code>, <code>species_id</code>, and <code>sex</code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"> yearly_sex_counts &lt;-<span class="st"> </span>surveys_complete <span class="op">%&gt;%</span>
<span class="st">                      </span><span class="kw">count</span>(year, species_id, sex)</code></pre></div>
<p>We can now make the faceted plot by splitting further by sex using <code>color</code> (within a single plot):</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"> <span class="kw">ggplot</span>(<span class="dt">data =</span> yearly_sex_counts, <span class="dt">mapping =</span> <span class="kw">aes</span>(<span class="dt">x =</span> year, <span class="dt">y =</span> n, <span class="dt">color =</span> sex)) <span class="op">+</span>
<span class="st">     </span><span class="kw">geom_line</span>() <span class="op">+</span>
<span class="st">     </span><span class="kw">facet_wrap</span>(<span class="op">~</span><span class="st"> </span>species_id)</code></pre></div>
<p><img src="WSBIM1207_files/figure-html/facet-by-species-and-sex-1.png" width="672"  /></p>
<p>Usually plots with white background look more readable when printed. We can set the background to white using the function <code>theme_bw()</code>. Additionally, you can remove the grid:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"> <span class="kw">ggplot</span>(<span class="dt">data =</span> yearly_sex_counts, <span class="dt">mapping =</span> <span class="kw">aes</span>(<span class="dt">x =</span> year, <span class="dt">y =</span> n, <span class="dt">color =</span> sex)) <span class="op">+</span>
<span class="st">     </span><span class="kw">geom_line</span>() <span class="op">+</span>
<span class="st">     </span><span class="kw">facet_wrap</span>(<span class="op">~</span><span class="st"> </span>species_id) <span class="op">+</span>
<span class="st">     </span><span class="kw">theme_bw</span>() <span class="op">+</span>
<span class="st">     </span><span class="kw">theme</span>(<span class="dt">panel.grid =</span> <span class="kw">element_blank</span>())</code></pre></div>
<p><img src="WSBIM1207_files/figure-html/facet-by-species-and-sex-white-bg-1.png" width="672"  /></p>
</div>
<div id="ggplot2-themes" class="section level2">
<h2><span class="header-section-number">5.6</span> <strong><code>ggplot2</code></strong> themes</h2>
<p>In addition to <code>theme_bw()</code>, which changes the plot background to white, <strong><code>ggplot2</code></strong> comes with several other themes which can be useful to quickly change the look of your visualization. The complete list of themes is available at <a href="http://docs.ggplot2.org/current/ggtheme.html" class="uri">http://docs.ggplot2.org/current/ggtheme.html</a>. <code>theme_minimal()</code> and <code>theme_light()</code> are popular, and <code>theme_void()</code> can be useful as a starting point to create a new hand-crafted theme.</p>
<p>The <a href="https://jrnold.github.io/ggthemes/reference/index.html">ggthemes</a> package provides a wide variety of options (including an Excel 2003 theme). The <a href="https://www.ggplot2-exts.org"><strong><code>ggplot2</code></strong> extensions website</a> provides a list of packages that extend the capabilities of <strong><code>ggplot2</code></strong>, including additional themes.</p>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Use what you just learned to create a plot that depicts how the average weight of each species changes through the years.</p>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<div class="solution">
<p class="solution-begin">
► Solution<span id="sol-start-39" class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility(&#39;sol-body-39&#39;, &#39;sol-start-39&#39;)"></span>
</p>
<div id="sol-body-39" class="solution-body" style="display: none;">
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">yearly_weight &lt;-<span class="st"> </span>surveys_complete <span class="op">%&gt;%</span>
<span class="st">    </span><span class="kw">group_by</span>(year, species_id) <span class="op">%&gt;%</span>
<span class="st">    </span><span class="kw">summarize</span>(<span class="dt">avg_weight =</span> <span class="kw">mean</span>(weight))
<span class="kw">ggplot</span>(<span class="dt">data =</span> yearly_weight, <span class="dt">mapping =</span> <span class="kw">aes</span>(<span class="dt">x=</span>year, <span class="dt">y=</span>avg_weight)) <span class="op">+</span>
<span class="st">    </span><span class="kw">geom_line</span>() <span class="op">+</span>
<span class="st">    </span><span class="kw">facet_wrap</span>(<span class="op">~</span><span class="st"> </span>species_id) <span class="op">+</span>
<span class="st">    </span><span class="kw">theme_bw</span>()</code></pre></div>
<p><img src="WSBIM1207_files/figure-html/average-weight-time-series-1.png" width="672"  /></p>
<p class="solution-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<p>The <code>facet_wrap</code> geometry extracts plots into an arbitrary number of dimensions to allow them to cleanly fit on one page. On the other hand, the <code>facet_grid</code> geometry allows you to explicitly specify how you want your plots to be arranged via formula notation (<code>rows ~ columns</code>; a <code>.</code> can be used as a placeholder that indicates only one row or column).</p>
<p>Let’s modify the previous plot to compare how the weights of males and females has changed through time:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># One column, facet by rows</span>
yearly_sex_weight &lt;-<span class="st"> </span>surveys_complete <span class="op">%&gt;%</span>
<span class="st">    </span><span class="kw">group_by</span>(year, sex, species_id) <span class="op">%&gt;%</span>
<span class="st">    </span><span class="kw">summarize</span>(<span class="dt">avg_weight =</span> <span class="kw">mean</span>(weight))
<span class="kw">ggplot</span>(<span class="dt">data =</span> yearly_sex_weight,
       <span class="dt">mapping =</span> <span class="kw">aes</span>(<span class="dt">x =</span> year, <span class="dt">y =</span> avg_weight, <span class="dt">color =</span> species_id)) <span class="op">+</span>
<span class="st">    </span><span class="kw">geom_line</span>() <span class="op">+</span>
<span class="st">    </span><span class="kw">facet_grid</span>(sex <span class="op">~</span><span class="st"> </span>.)</code></pre></div>
<p><img src="WSBIM1207_files/figure-html/average-weight-time-facet-sex-rows-1.png" width="672"  /></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># One row, facet by column</span>
<span class="kw">ggplot</span>(<span class="dt">data =</span> yearly_sex_weight,
       <span class="dt">mapping =</span> <span class="kw">aes</span>(<span class="dt">x =</span> year, <span class="dt">y =</span> avg_weight, <span class="dt">color =</span> species_id)) <span class="op">+</span>
<span class="st">    </span><span class="kw">geom_line</span>() <span class="op">+</span>
<span class="st">    </span><span class="kw">facet_grid</span>(. <span class="op">~</span><span class="st"> </span>sex)</code></pre></div>
<p><img src="WSBIM1207_files/figure-html/average-weight-time-facet-sex-columns-1.png" width="672"  /></p>
</div>
<div id="customization" class="section level2">
<h2><span class="header-section-number">5.7</span> Customization</h2>
<p>Take a look at the <a href="https://www.rstudio.com/wp-content/uploads/2016/11/ggplot2-cheatsheet-2.1.pdf"><strong><code>ggplot2</code></strong> cheat sheet</a>, and think of ways you could improve the plot.</p>
<p>Now, let’s change names of axes to something more informative than ‘year’ and ‘n’ and add a title to the figure:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">ggplot</span>(<span class="dt">data =</span> yearly_sex_counts, <span class="dt">mapping =</span> <span class="kw">aes</span>(<span class="dt">x =</span> year, <span class="dt">y =</span> n, <span class="dt">color =</span> sex)) <span class="op">+</span>
<span class="st">    </span><span class="kw">geom_line</span>() <span class="op">+</span>
<span class="st">    </span><span class="kw">facet_wrap</span>(<span class="op">~</span><span class="st"> </span>species_id) <span class="op">+</span>
<span class="st">    </span><span class="kw">labs</span>(<span class="dt">title =</span> <span class="st">&quot;Observed species in time&quot;</span>,
         <span class="dt">x =</span> <span class="st">&quot;Year of observation&quot;</span>,
         <span class="dt">y =</span> <span class="st">&quot;Number of species&quot;</span>) <span class="op">+</span>
<span class="st">    </span><span class="kw">theme_bw</span>()</code></pre></div>
<p><img src="WSBIM1207_files/figure-html/number-species-year-with-right-labels-1.png" width="672"  /></p>
<p>The axes have more informative names, but their readability can be improved by increasing the font size:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">ggplot</span>(<span class="dt">data =</span> yearly_sex_counts, <span class="dt">mapping =</span> <span class="kw">aes</span>(<span class="dt">x =</span> year, <span class="dt">y =</span> n, <span class="dt">color =</span> sex)) <span class="op">+</span>
<span class="st">    </span><span class="kw">geom_line</span>() <span class="op">+</span>
<span class="st">    </span><span class="kw">facet_wrap</span>(<span class="op">~</span><span class="st"> </span>species_id) <span class="op">+</span>
<span class="st">    </span><span class="kw">labs</span>(<span class="dt">title =</span> <span class="st">&quot;Observed species in time&quot;</span>,
        <span class="dt">x =</span> <span class="st">&quot;Year of observation&quot;</span>,
        <span class="dt">y =</span> <span class="st">&quot;Number of species&quot;</span>) <span class="op">+</span>
<span class="st">    </span><span class="kw">theme_bw</span>() <span class="op">+</span>
<span class="st">    </span><span class="kw">theme</span>(<span class="dt">text=</span><span class="kw">element_text</span>(<span class="dt">size =</span> <span class="dv">16</span>))</code></pre></div>
<p><img src="WSBIM1207_files/figure-html/number-species-year-with-right-labels-xfont-size-1.png" width="672"  /></p>
<p>Note that it is also possible to change the fonts of your plots. If you are on Windows, you may have to install the <a href="https://github.com/wch/extrafont"><strong><code>extrafont</code></strong> package</a>, and follow the instructions included in the README for this package.</p>
<p>After our manipulations, you may notice that the values on the x-axis are still not properly readable. Let’s change the orientation of the labels and adjust them vertically and horizontally so they don’t overlap. You can use a 90-degree angle, or experiment to find the appropriate angle for diagonally oriented labels:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">ggplot</span>(<span class="dt">data =</span> yearly_sex_counts, <span class="dt">mapping =</span> <span class="kw">aes</span>(<span class="dt">x =</span> year, <span class="dt">y =</span> n, <span class="dt">color =</span> sex)) <span class="op">+</span>
<span class="st">    </span><span class="kw">geom_line</span>() <span class="op">+</span>
<span class="st">    </span><span class="kw">facet_wrap</span>(<span class="op">~</span><span class="st"> </span>species_id) <span class="op">+</span>
<span class="st">    </span><span class="kw">labs</span>(<span class="dt">title =</span> <span class="st">&quot;Observed species in time&quot;</span>,
        <span class="dt">x =</span> <span class="st">&quot;Year of observation&quot;</span>,
        <span class="dt">y =</span> <span class="st">&quot;Number of species&quot;</span>) <span class="op">+</span>
<span class="st">    </span><span class="kw">theme_bw</span>() <span class="op">+</span>
<span class="st">    </span><span class="kw">theme</span>(<span class="dt">axis.text.x =</span> <span class="kw">element_text</span>(<span class="dt">colour =</span> <span class="st">&quot;grey20&quot;</span>, <span class="dt">size =</span> <span class="dv">12</span>, <span class="dt">angle =</span> <span class="dv">90</span>, <span class="dt">hjust =</span> <span class="fl">0.5</span>, <span class="dt">vjust =</span> <span class="fl">0.5</span>),
                        <span class="dt">axis.text.y =</span> <span class="kw">element_text</span>(<span class="dt">colour =</span> <span class="st">&quot;grey20&quot;</span>, <span class="dt">size =</span> <span class="dv">12</span>),
          <span class="dt">text =</span> <span class="kw">element_text</span>(<span class="dt">size =</span> <span class="dv">16</span>))</code></pre></div>
<p><img src="WSBIM1207_files/figure-html/number-species-year-with-theme-1.png" width="672"  /></p>
<p>If you like the changes you created better than the default theme, you can save them as an object to be able to easily apply them to other plots you may create:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">grey_theme &lt;-<span class="st"> </span><span class="kw">theme</span>(<span class="dt">axis.text.x =</span> <span class="kw">element_text</span>(<span class="dt">colour =</span> <span class="st">&quot;grey20&quot;</span>, <span class="dt">size =</span> <span class="dv">12</span>, <span class="dt">angle =</span> <span class="dv">90</span>, <span class="dt">hjust =</span> <span class="fl">0.5</span>, <span class="dt">vjust =</span> <span class="fl">0.5</span>),
                          <span class="dt">axis.text.y =</span> <span class="kw">element_text</span>(<span class="dt">colour =</span> <span class="st">&quot;grey20&quot;</span>, <span class="dt">size =</span> <span class="dv">12</span>),
                          <span class="dt">text =</span> <span class="kw">element_text</span>(<span class="dt">size =</span> <span class="dv">16</span>))
<span class="kw">ggplot</span>(surveys_complete, <span class="kw">aes</span>(<span class="dt">x =</span> species_id, <span class="dt">y =</span> hindfoot_length)) <span class="op">+</span>
<span class="st">    </span><span class="kw">geom_boxplot</span>() <span class="op">+</span>
<span class="st">    </span>grey_theme</code></pre></div>
<p><img src="WSBIM1207_files/figure-html/number-species-year-with-right-labels-xfont-orientation-1.png" width="672"  /></p>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>With all of this information in hand, please take another five minutes to either improve one of the plots generated in this exercise or create a beautiful graph of your own. Use the RStudio <a href="https://www.rstudio.com/wp-content/uploads/2016/11/ggplot2-cheatsheet-2.1.pdf"><strong><code>ggplot2</code></strong> cheat sheet</a> for inspiration. Here are some ideas:</p>
<ul>
<li>See if you can change the thickness of the lines.</li>
<li>Can you find a way to change the name of the legend? What about its labels?</li>
<li>Try using a different color palette (see <a href="http://www.cookbook-r.com/Graphs/Colors_(ggplot2)/" class="uri">http://www.cookbook-r.com/Graphs/Colors_(ggplot2)/</a>).</li>
</ul>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
</div>
<div id="arranging-and-exporting-plots" class="section level2">
<h2><span class="header-section-number">5.8</span> Arranging and exporting plots</h2>
<p>Faceting is a great tool for splitting one plot into multiple plots, but sometimes you may want to produce a single figure that contains multiple plots using different variables or even different data frames. The <strong><code>gridExtra</code></strong> package allows us to combine separate ggplots into a single figure using <code>grid.arrange()</code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">install.packages</span>(<span class="st">&quot;gridExtra&quot;</span>)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(gridExtra)

spp_weight_boxplot &lt;-<span class="st"> </span><span class="kw">ggplot</span>(<span class="dt">data =</span> surveys_complete,
                             <span class="dt">mapping =</span> <span class="kw">aes</span>(<span class="dt">x =</span> species_id, <span class="dt">y =</span> weight)) <span class="op">+</span>
<span class="st">  </span><span class="kw">geom_boxplot</span>() <span class="op">+</span>
<span class="st">  </span><span class="kw">xlab</span>(<span class="st">&quot;Species&quot;</span>) <span class="op">+</span><span class="st"> </span><span class="kw">ylab</span>(<span class="st">&quot;Weight (g)&quot;</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw">scale_y_log10</span>()

spp_count_plot &lt;-<span class="st"> </span><span class="kw">ggplot</span>(<span class="dt">data =</span> yearly_counts,
                         <span class="dt">mapping =</span> <span class="kw">aes</span>(<span class="dt">x =</span> year, <span class="dt">y =</span> n, <span class="dt">color =</span> species_id)) <span class="op">+</span>
<span class="st">  </span><span class="kw">geom_line</span>() <span class="op">+</span>
<span class="st">  </span><span class="kw">xlab</span>(<span class="st">&quot;Year&quot;</span>) <span class="op">+</span><span class="st"> </span><span class="kw">ylab</span>(<span class="st">&quot;Abundance&quot;</span>)

<span class="kw">grid.arrange</span>(spp_weight_boxplot, spp_count_plot, <span class="dt">ncol =</span> <span class="dv">2</span>, <span class="dt">widths =</span> <span class="kw">c</span>(<span class="dv">4</span>, <span class="dv">6</span>))</code></pre></div>
<p><img src="WSBIM1207_files/figure-html/gridarrange-example-1.png" width="960"  /></p>
<p>In addition to the <code>ncol</code> and <code>nrow</code> arguments, used to make simple arrangements, there are tools for <a href="https://cran.r-project.org/web/packages/gridExtra/vignettes/arrangeGrob.html">constructing more complex layouts</a>.</p>
<p>After creating your plot, you can save it to a file in your favorite format. The Export tab in the <strong>Plot</strong> pane in RStudio will save your plots at low resolution, which will not be accepted by many journals and will not scale well for posters.</p>
<p>Instead, use the <code>ggsave()</code> function, which allows you easily change the dimension and resolution of your plot by adjusting the appropriate arguments (<code>width</code>, <code>height</code> and <code>dpi</code>).</p>
<p>Make sure you have the <code>fig_output/</code> folder in your working directory.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">my_plot &lt;-<span class="st"> </span><span class="kw">ggplot</span>(<span class="dt">data =</span> yearly_sex_counts,
                  <span class="dt">mapping =</span> <span class="kw">aes</span>(<span class="dt">x =</span> year, <span class="dt">y =</span> n, <span class="dt">color =</span> sex)) <span class="op">+</span>
<span class="st">    </span><span class="kw">geom_line</span>() <span class="op">+</span>
<span class="st">    </span><span class="kw">facet_wrap</span>(<span class="op">~</span><span class="st"> </span>species_id) <span class="op">+</span>
<span class="st">    </span><span class="kw">labs</span>(<span class="dt">title =</span> <span class="st">&quot;Observed species in time&quot;</span>,
        <span class="dt">x =</span> <span class="st">&quot;Year of observation&quot;</span>,
        <span class="dt">y =</span> <span class="st">&quot;Number of species&quot;</span>) <span class="op">+</span>
<span class="st">    </span><span class="kw">theme_bw</span>() <span class="op">+</span>
<span class="st">    </span><span class="kw">theme</span>(<span class="dt">axis.text.x =</span> <span class="kw">element_text</span>(<span class="dt">colour =</span> <span class="st">&quot;grey20&quot;</span>, <span class="dt">size =</span> <span class="dv">12</span>, <span class="dt">angle =</span> <span class="dv">90</span>, <span class="dt">hjust =</span> <span class="fl">0.5</span>, <span class="dt">vjust =</span> <span class="fl">0.5</span>),
                        <span class="dt">axis.text.y =</span> <span class="kw">element_text</span>(<span class="dt">colour =</span> <span class="st">&quot;grey20&quot;</span>, <span class="dt">size =</span> <span class="dv">12</span>),
          <span class="dt">text=</span><span class="kw">element_text</span>(<span class="dt">size =</span> <span class="dv">16</span>))
<span class="kw">ggsave</span>(<span class="st">&quot;fig_output/yearly_sex_counts.png&quot;</span>, my_plot, <span class="dt">width =</span> <span class="dv">15</span>, <span class="dt">height =</span> <span class="dv">10</span>)

<span class="co"># This also works for grid.arrange() plots</span>
combo_plot &lt;-<span class="st"> </span><span class="kw">grid.arrange</span>(spp_weight_boxplot, spp_count_plot, <span class="dt">ncol =</span> <span class="dv">2</span>, <span class="dt">widths =</span> <span class="kw">c</span>(<span class="dv">4</span>, <span class="dv">6</span>))
<span class="kw">ggsave</span>(<span class="st">&quot;fig_output/combo_plot_abun_weight.png&quot;</span>, combo_plot, <span class="dt">width =</span> <span class="dv">10</span>, <span class="dt">dpi =</span> <span class="dv">300</span>)</code></pre></div>
<p>Note: The parameters <code>width</code> and <code>height</code> also determine the font size in the saved plot.</p>
</div>
<div id="other-packages-for-visualisation" class="section level2">
<h2><span class="header-section-number">5.9</span> Other packages for visualisation</h2>
<p><code>ggplot2</code> is a very powerful package that fits very nicely in our <em>tidy data</em> and <em>tidy tools</em> pipeline. There are other visualisation packages in R that shouldn’t be ignored.</p>
<div id="base-graphics" class="section level3 unnumbered">
<h3>Base graphics</h3>
<p>The default graphics system that comes with R, often called <em>base R graphics</em> is simple and fast. It is based on the <em>painters model</em>, where different output are directly overlaid on top of each other (see figure <a href="sec-vis.html#fig:paintermodel">5.1</a>). This is a fundamental difference with <code>ggplot2</code> (and with <code>lattice</code>, described below), that returns dedicated objects, that are rendered on screen or in a file, and that can even be updated.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">par</span>(<span class="dt">mfrow =</span> <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">3</span>))
<span class="kw">plot</span>(<span class="dv">1</span><span class="op">:</span><span class="dv">20</span>, <span class="dt">main =</span> <span class="st">&quot;First layer, produced with plot(1:20)&quot;</span>)
<span class="kw">plot</span>(<span class="dv">1</span><span class="op">:</span><span class="dv">20</span>, <span class="dt">main =</span> <span class="st">&quot;A horizontal red line, added with abline(h = 10)&quot;</span>)
<span class="kw">abline</span>(<span class="dt">h =</span> <span class="dv">10</span>, <span class="dt">col =</span> <span class="st">&quot;red&quot;</span>)
<span class="kw">plot</span>(<span class="dv">1</span><span class="op">:</span><span class="dv">20</span>, <span class="dt">main =</span> <span class="st">&quot;A rectangle, added with rect(5, 5, 15, 15)&quot;</span>)
<span class="kw">abline</span>(<span class="dt">h =</span> <span class="dv">10</span>, <span class="dt">col =</span> <span class="st">&quot;red&quot;</span>)
<span class="kw">rect</span>(<span class="dv">5</span>, <span class="dv">5</span>, <span class="dv">15</span>, <span class="dv">15</span>, <span class="dt">lwd =</span> <span class="dv">3</span>)</code></pre></div>
<div class="figure"><span id="fig:paintermodel"></span>
<p class="caption marginnote shownote">
Figure 5.1: Successive layers added on top of each other.
</p>
<img src="WSBIM1207_files/figure-html/paintermodel-1.png" alt="Successive layers added on top of each other." width="1152"  />
</div>
<p>Another main difference is that base graphics’ plotting function try to do <em>the right</em> thing based on their input type, i.e. they will adapt their behaviour based on the class of their input. This is again very different from what we have in <code>ggplot2</code>, that only accepts dataframes as input, and that requires plots to be constructed bit by bit.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">par</span>(<span class="dt">mfrow =</span> <span class="kw">c</span>(<span class="dv">2</span>, <span class="dv">2</span>))
<span class="kw">boxplot</span>(<span class="kw">rnorm</span>(<span class="dv">100</span>),
        <span class="dt">main =</span> <span class="st">&quot;Boxplot of rnorm(100)&quot;</span>)
<span class="kw">boxplot</span>(<span class="kw">matrix</span>(<span class="kw">rnorm</span>(<span class="dv">100</span>), <span class="dt">ncol =</span> <span class="dv">10</span>),
        <span class="dt">main =</span> <span class="st">&quot;Boxplot of matrix(rnorm(100), ncol = 10)&quot;</span>)
<span class="kw">hist</span>(<span class="kw">rnorm</span>(<span class="dv">100</span>))
<span class="kw">hist</span>(<span class="kw">matrix</span>(<span class="kw">rnorm</span>(<span class="dv">100</span>), <span class="dt">ncol =</span> <span class="dv">10</span>))</code></pre></div>
<div class="figure"><span id="fig:plotmethod"></span>
<p class="caption marginnote shownote">
Figure 5.2: Plotting boxplots (top) and histograms (bottom) vectors (left) or a matrices (right).
</p>
<img src="WSBIM1207_files/figure-html/plotmethod-1.png" alt="Plotting boxplots (top) and histograms (bottom) vectors (left) or a matrices (right)." width="768"  />
</div>
<p>The out-of-the-box approach in base graphics can be very efficient for simple, standard figures, that can be produced very quickly with a single line of code and a single function such as <code>plot</code>, or <code>hist</code>, or <code>boxplot</code>, … The defaults are however not always the most appealing and tuning of figures, especially when they become more complex (for example to produce facets), can become lengthy and cumbersome.</p>
</div>
<div id="the-lattice-package" class="section level3 unnumbered">
<h3>The lattice package</h3>
<p>The <code>lattice</code> package is similar to <code>ggplot2</code> in that is uses dataframes as input, returns graphical objects and supports faceting. <code>lattice</code> however isn’t based on the grammar of graphics and has a more convoluted interface.</p>
<p>A good reference for the <code>lattice</code> package is <span class="citation">Deepayan (<label for="tufte-mn-21" class="margin-toggle">&#8853;</label><input type="checkbox" id="tufte-mn-21" class="margin-toggle">2008<span class="marginnote">Deepayan, Sarkar. 2008. <em>Lattice: Multivariate Data Visualization with R</em>. New York: Springer. <a href="http://lmdvr.r-forge.r-project.org" class="uri">http://lmdvr.r-forge.r-project.org</a>.</span>)</span>.</p>
</div>
</div>
<div id="additional-exercises-4" class="section level2">
<h2><span class="header-section-number">5.10</span> Additional exercises</h2>
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>Load the beer consuption data with</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(<span class="st">&quot;rWSBIM1207&quot;</span>)
<span class="kw">data</span>(beers)</code></pre></div>
<p>Analyse the data to answer the question <em>Do men drink more than women according to age and working status?</em> Now reproduce the figure below.</p>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-181"></span>
<p class="caption marginnote shownote">
Figure 5.3: Visualisation of beer consumption, highlighting different patterns of beer consumption in employed and unemployed males and females.
</p>
<img src="figs/ggplot-beer.png" alt="Visualisation of beer consumption, highlighting different patterns of beer consumption in employed and unemployed males and females." width="100%"  />
</div>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<!-- 

<div class = "solution"><p class="solution-begin">&#x25BA; Solution<span id='sol-start-40' class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-40', 'sol-start-40')"></span></p><div class="solution-body" id = "sol-body-40" style="display: none;">

 -->
<!-- ```{r} -->
<!-- library("dplyr") -->
<!-- beers3 <- beers %>% -->
<!--     filter(!is.na(Consumption)) %>% -->
<!--     group_by(Gender, Age, Work) %>% -->
<!--     summarize(mean_consumption = mean(Consumption)) -->
<!-- library("ggplot2") -->
<!-- ggplot(data = beers3, -->
<!--        mapping = aes(x = Age, y = mean_consumption, colour = Work)) + -->
<!--   geom_line() + -->
<!--   geom_point() + -->
<!--   facet_wrap(~ Gender) + -->
<!--   labs(title = "Mean consumption of beers according to age, employment status and gender", -->
<!--        y = "Mean consumption (Litre)") -->
<!-- ``` -->
<!-- 

<p class="solution-end"><span class="fa fa-square-o solution-icon"></span></p></div></div>

 -->
<div class="question">
<p class="question-begin">
► Question
</p>
<div class="question-body">
<p>We are now going to analyse transcriptomics data from the TCGA project.</p>
<ul>
<li><p>Using the file returned by the <code>expression.csv()</code> function from the <code>rWSBIM1207</code> package, create a table called <code>expression</code>.</p></li>
<li><p>Inspect the data.</p></li>
</ul>
<p>The <code>sampleID</code> column gives the TCGA reference of the sample and sample is unique. The 12 first characters (TCGA-XX-XXXX) of the samples ID are unique to each patient. The 3 last characters of the samples ID indicate the type of sample: samples ending with ‘-01A’ (TCGA-XX-XXXX-01A) correspond to tumors and samples ending with ‘-11A’ (TCGA-XX-XXXX-11A) correspond to normal peritumoral tissues.</p>
<p>The <code>patient</code> column gives the reference of each patient. Note that some patient for which both tumor and normal tissue were analysed are recorded twice. The <code>type</code> column gives the nature of each sample (tumor or normal tissue). The 5 next column give the expression levels of 5 genes in each sample.</p>
<ul>
<li><p>Using geom_point, draw a plot showing distribution of expression levels of A2LD1 in normal tissue samples and in primary tumor samples.</p></li>
<li><p>Repeat this visualisation using this time the geom_jitter. Which representation is more appropriate? Why?</p></li>
<li><p>Colour the samples according to the tissue type.</p></li>
<li><p>Colour the samples according to their expression level in A2ML1.</p></li>
<li><p>Highlight the points corresponding to patient “TCGA-73-4676”.</p></li>
<li><p>Add a transparent boxplot to the graph.</p></li>
<li><p>Change the y scale to log10 scale.</p></li>
<li><p>Visualise the expression of A1BG against that of A2LD1 setting the x axis on the log10 scale. Colours the observations based on their type and resize the points according to the expression level of the A2ML1 gene.</p></li>
</ul>
<p class="question-end">
<span class="fa fa-square-o solution-icon"></span>
</p>
</div>
</div>
<!-- 

<div class = "solution"><p class="solution-begin">&#x25BA; Solution<span id='sol-start-41' class="fa fa-plus-square solution-icon clickable" onclick="toggle_visibility('sol-body-41', 'sol-start-41')"></span></p><div class="solution-body" id = "sol-body-41" style="display: none;">

 -->
<!-- ```{r} -->
<!-- library("rWSBIM1207") -->
<!-- library("readr") -->
<!-- library("ggplot2") -->
<!-- expression <- read_csv(expression.csv()) -->
<!-- ## Using geom_point, draw a plot showing distribution of expression -->
<!-- ## levels of A2LD1 in normal tissue samples and in primary tumor samples -->
<!-- ggplot(data = expression, -->
<!--        mapping = aes(x = type, y = A2LD1)) + -->
<!--   geom_point() -->
<!-- ## Repeat this visualisation using this time the geom_jitter. Which -->
<!-- ## representation is more appropriate? Why? -->
<!-- ggplot(data = expression, -->
<!--        mapping = aes(x = type, y = A2LD1)) + -->
<!--     geom_jitter() -->
<!-- ## Colour the samples according to the tissue type. -->
<!-- ggplot(data = expression, -->
<!--        mapping = aes(x = type, y = A2LD1, colour = type)) + -->
<!--   geom_jitter() -->
<!-- ## Colour the samples according to their expression level in A2ML1. -->
<!-- ggplot(data = expression, -->
<!--        mapping = aes(x = type, y = A2LD1, colour = A2ML1)) + -->
<!--   geom_jitter() -->
<!-- ## Highlight the points corresponding to patient "TCGA-73-4676" -->
<!-- ggplot(data = expression, -->
<!--        mapping = aes(x = type, y = A2LD1, -->
<!--                      colour = patient == "TCGA-73-4676")) + -->
<!--   geom_jitter() -->
<!-- ## Add a transparent boxplot to the graph. -->
<!-- ggplot(data = expression, mapping = aes(x = type, y = A2LD1)) + -->
<!--   geom_jitter(aes(colour = type)) + -->
<!--   geom_boxplot(alpha = 0) -->
<!-- ## Change the y scale to log10 scale. -->
<!-- ggplot(data = expression, mapping = aes(x = type, y = A2LD1)) + -->
<!--   geom_jitter(aes(colour = type)) + -->
<!--   geom_boxplot(alpha = 0) + -->
<!--     scale_y_log10() -->
<!-- ## Visualise the expression of A1BG against that of A2LD1 setting the -->
<!-- ## x axis on the log10 scale. Colours the observations based on their -->
<!-- ## type and resize the points according to the expression level of the -->
<!-- ## A2ML1 gene. -->
<!-- ggplot(data = expression, -->
<!--        mapping = aes(x = A1BG, y = A2LD1, -->
<!--                      colour = type, -->
<!--                      size = A2ML1)) + -->
<!--     geom_point() + -->
<!--     scale_x_log10() -->
<!-- ``` -->
<!-- 

<p class="solution-end"><span class="fa fa-square-o solution-icon"></span></p></div></div>

 -->

</div>
</div>
<p style="text-align: center;">
<a href="sec-dplyr.html"><button class="btn btn-default">Previous</button></a>
<a href="sec-join.html"><button class="btn btn-default">Next</button></a>
</p>
<p class="build-date">Page built: 
2019-03-12
</p>
</div>
</div>



</body>
</html>
